app/services/letter_sending_service.rb

```diff
@@ -3,27 +3,31 @@
 require "notifications/client"
 
 class LetterSendingService
-  attr_reader :neighbour, :consultation, :letter_content, :resend_reason
+  attr_reader :consultation, :letter_content, :resend_reason
 
-  def initialize(neighbour, letter_content, letter_type:, resend_reason: nil)
-    @local_authority = neighbour.consultation.planning_application.local_authority
-    @neighbour = neighbour
-    @consultation = neighbour.consultation
+  def initialize(letter_content, letter_type:, resend_reason: nil)
     @letter_content = letter_content
     @resend_reason = resend_reason
     @letter_type = letter_type
   end
 
-  def deliver!
+  def deliver!(neighbour)
```

Comment:
We have quite a lot of updates / db transactions happening in this method. Is it worth putting everything in the ActiveRecord::Base.transaction in case we need to rollback if one step fails?

Reply:
Think we can also add more specific error handling i.e `ActiveRecord::RecordInvalid` to this method

Reply:
```ruby
def deliver!(neighbour)
  return if resend_reason.nil? && NeighbourLetter.find_by(neighbour:).present? && consultation_letter?

  @local_authority = neighbour.consultation.planning_application.local_authority
  @consultation = neighbour.consultation
  letter_record = NeighbourLetter.new(neighbour:, text: letter_content, resend_reason:)

  ActiveRecord::Base.transaction do
    begin
      letter_record.save!
      consultation.start_deadline if consultation_letter?

      @batch.neighbour_letters << letter_record if @batch

      if resend_reason.present?
        letter_content.prepend "# Application updated\nThis application has been updated. Reason: #{resend_reason}\n\n"
      end

      personalisation = { message: letter_content, heading: }.merge(neighbour.format_address_lines)

      response = client.send_letter(
        template_id: @local_authority.letter_template_id,
        personalisation:
      )

      update_letter!(letter_record, response)
      neighbour.touch :last_letter_sent_at

    rescue ActiveRecord::RecordInvalid => e
      Appsignal.report_error(e)
    rescue Notifications::Client::RequestError => e
      letter_record.update!(status: "rejected", failure_reason: e.message)
      Appsignal.report_error(e)
    rescue StandardError => e
      # General fallback
      letter_record.update!(status: "rejected", failure_reason: e.message)
      Appsignal.report_error(e)
    end
  end
end

def deliver_batch!(neighbours)
  @batch = neighbours.first.consultation.neighbour_letter_batches.new(text: letter_content)
  neighbours.each do |neighbour|
    deliver!(neighbour)
  rescue => e
    Appsignal.report_error(e)
    next
  end
end

```

Reply:
Had a go at writing that method. Still seems like it's doing too much and perhaps we need to rollback in the rescue blocks if we don't want the partial updates going ahead?

Reply:
Wouldn't moving everything inside the transaction actually backfire slightly? It would mean, for example, that if a letter failed to send, it wouldn't get added to the batch, because that would get rolled back.

I'm also not sure it would be necessary to handle the errors in `deliver_batch!`, because every exception is caught inside `deliver!`.

Reply:
You're right though that I'm not completely sure about the error handling.